## Let's Program... in Python!

Recall Game Studies journal:

> ISSN:1604-7982

It is preferable to write as few lines per cell as possible.  That way you can "debug" your code in real time (JupyterLab is unreasonably magical about making that work).  For example:

Let's create a variable called ISSN and give it our ISSN number as a "string" (a word) not an integer.  If we treat it like an integer, 1604798 becomes 1,604,798, and that is irrelevant to us.  We want to look at each letter as though it exists in its own world. 

So we use an "assignment operator," also called an "equals sign";

and then in order to create a string (basically, a human-style word) we use quotes.

In [34]:
ISSN = "1604798"

That's it!  Now (unlike other programming environments) ISSN will hold onto that value for a while!

Let's try it out by asking it to "print" the value of ISSN.

In [35]:
print(ISSN)

1604798


See how it puts the output in the space below the line?  If we go back to those lines and run them again, the new output will reflect the new values (if there are any).

Because it is so common to ask the computer about the value of a variable, the shortcut is just to type out the name of the variable itself.

In [36]:
ISSN

'1604798'

One key idea:  Spaces are bad.  Avoid spaces whenever you can.  And to make things easier, always try to give variables names that will be easy to understand a month from now.  For example, reflect on the utility of these three variable names:

In [37]:
p = "Gwen"

In [38]:
person = "Gwen"

In [39]:
mySwornEnemy = "Gwen"

See how that works?  But back to our ISSN problem.

First:  Don't get hung up on using the tool "like it is supposed to be used."  There is no right way to use it.  Worst case scenario:  Just use JupyterLab like a big dumb calculator!  There's nothing wrong with that ("whatever works"), and it is still hella useful.  To wit:

In [41]:
total = (8*1) + (7*6) + (6*0) + (5*4) + (4*7) + (3*9) + (2*8)

In [42]:
print(total)

141


Then we get the modulo (cool kids say "mod," we abbreviate it with percent sign, 'cause it is related to division, and the percent sign is a fancy division sign).  So the total *modulo* 11.

In [45]:
total=total%11

In [46]:
print(total)

9


Last step:  Subtract that number from 11.

In [47]:
print(11-9)

2


Huzzah!  That's what we were looking for.  So that's one way to do it.  But we can write some code to do some (or all) of it, too.  Here's what that would look like.

*(Self-evidently, every line of code increases the likelihood that something will go wrong.  Just accept that, and carry on.)*

### first

get the length (number of characters) of our ISSN string, and then add 1 to that number.  My goal is to count down from the left to the right so the last digit (on the right) has a multiplier of 2 (see our calculations above).

In [50]:
digitMultiplier = len(ISSN) + 1

In [56]:
digitMultiplier

8

Remember that this is a shortcut to saying "print(digitMultiplier)"

Now it gets a bit more complicated, because I'm going to ask the computer to do something repeatedly and track the answer as it changes.  I'm going to ask the computer to "iterate" over these characters and do the math with each iteration, so I'll need a running total.  Let's create it ahead of time.

In [53]:
runningTotal = 0

In [65]:
runningTotal

0

Ready?  Cause it gets weird from this point forward.  I'll build it out slowly and test parts of the code as I go along -- that's how I write 90% of my Python code, because Python Is Not My Friend.

In [67]:
for eachCharacter in ISSN:
    print(eachCharacter)

1
6
0
4
7
9
8


OK, but those characters are not numbers, they are strings -- "6" instead of 6.  Watch how this fails (I'm adding "1" arbitrarily, just to show it failing/working):

In [68]:
for eachCharacter in ISSN:
    print(eachCharacter+1)

TypeError: can only concatenate str (not "int") to str

Basically, "concatenate" means "push letters together."  So when I have integers, 3+5+7=15; but when I'm using strings, 3+5+7=357.  Since "1604798" started as a string, each character it gets from that string is still a string.  So we need to "transform" each character into an integer.  We use the int() function, which -- whenever possible -- converts a string into an integer.

In [69]:
for eachCharacter in ISSN:
    myNumber = int(eachCharacter)
    print(myNumber+1)

2
7
1
5
8
10
9


*Cool.*

So now we have each number, one at a time, starting from the first digit in the series.  This is where that `digitMultiplier` variable from before comes in handy.  (Note that I have to keep retyping this part of the code because it is a "for" loop, and therefore exists as a kind of unit.

In [70]:
for eachCharacter in ISSN:
    myNumber = int(eachCharacter)
    myProduct = myNumber * digitMultiplier
    print(myProduct)

8
48
0
32
56
72
64


That's a start, but it isn't quite right:  Remember we need to DECREMENT each subsequent multiplier by one (8, then 7, then 6.... and eventually 2).  We're going to be changing that value (`digitMultiplier`) permanently along the way, so let's refresh the correct value by adding it to the top of the codepile (a technical term).

In [72]:
digitMultiplier = len(ISSN) + 1

In [73]:
myTotal = 0

In [74]:
for eachCharacter in ISSN:
    myNumber = int(eachCharacter)
    myProduct = myNumber * digitMultiplier
    # calculate a running total:
    myTotal = myTotal + myProduct
    # gradually reduce the value of the multiplier
    digitMultiplier = digitMultiplier - 1

What did we end up with?  Let's find out -- based on earlier results, it should be 141.

In [75]:
print(myTotal)

141


OMG IT WORKED.  It is gravy from this point forward:

In [76]:
myTotal = myTotal%11

In [77]:
myResult = 11-myTotal

In [78]:
print(myResult)

2


Remember way back when?  Check out the last digit in the sequence!

##  ISSN:1604-7982